bot/cogs/tags.py

```diff
@@ -86,11 +86,62 @@ def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None
             return self._get_suggestions(tag_name)
         return found
 
+    async def _get_tags_via_content(self, check: Callable[[Iterable], bool], keywords: str) -> Optional[Embed]:
+        """
+        Search for tags via contents.
+
+        `predicate` will be either any or all, or a custom callable to search. Must return a bool.
+        """
+        await self._get_tags()
+
+        keywords_processed: Tuple[str] = tuple(query.strip().casefold() for query in keywords.split(',') if query)
+        keywords_processed = keywords_processed or (keywords,)
+        founds: list = [
+            tag
+            for tag in self._cache.values()
+            if check(query in tag['embed']['description'] for query in keywords_processed)
+        ]
```

Comment:
This whole block is incredibly dense. I have a number of objections to it and think you should simplify and refactor the entire block.

- Why are you using type annotations here? It's perfectly obvious that the first line is a tuple from the `tuple()` call, and that the last line is a list from the fact that it's wrapped in square brackets. Please only use inline type annotations in cases where they add clarity, such as surprising types and cases in which the reader would otherwise have a hard time knowing what the type of some variable is.

- Please don't make list comprehensions with this much complexity. If your list comprehension has more than two expressions in it, just refactor it into regular code. See Brett Slatkins excellent book **Effective Python**, Chapter 1 Item 8 - "Avoid More Than Two Expressions in List Comprehensions" for more information. Not only does this have three expressions in it, but you've had to split it up onto multiple lines to even fit it. If you can't fit your list comprehension on a single line, this is usually a sign that you should consider refactoring.

- The `keywords_processed` lines are quite dense. Think about all the stuff you're doing in two lines here. Let's write it down in plain English: 

  Split the `keyword` variable by comma. Create a tuple of the results this produces. Every result should be stripped, and also casefolded. However, if the result is false-y, do not put it in the tuple. Once this process is done, if this process produces no results at all, replace this tuple with a tuple containing only a single item; whatever is in `keywords`.

  Not only that, but you're using `typing.Tuple`, `tuple()` and `(keywords,)` all in two lines. Imagine how confusing this would be to a beginner. Please try to simplify this part into something a beginner wouldn't be so scared by.

- This is the most complex part of this entire pull request, but there are no comments. Use block comments to explain complexity if you can't refactor it into simpler code, and otherwise, use block comments as a sort of bookmark to describe the block below.

Reply:
Thank you for very detailed reviews!

It does hit me now that you say "If you can't fit your list comprehension on a single line, this is usually a sign that you should consider refactoring." - it is a very good point, and I will definitely refactor it as well as the other dense part of the code.